core/src/main/java/org/apache/accumulo/core/spi/scan/DefaultScanServerDispatcher.java

```diff
@@ -40,53 +42,241 @@
  * </p>
  *
+ * The json is structured as a list of profiles, with each profile having the following fields.
+ *
+ * <ul>
+ * <li><b>isDefault : </b> A boolean that specifies whether this is the default profile. One and
+ * only one profile must set this to true.</li>
+ * <li><b>maxBusyTimeout : </b> The maximum busy timeout to use. The busy timeout from the last
+ * attempt configuration grows exponentially up to this max.</li>
+ * <li><b>scanTypeActivations : </b> A list of scan types that will activate this profile. Scan
+ * types are specified by setting {@code scan_type=<scan_type>} as execution on the scanner. See
+ * {@link org.apache.accumulo.core.client.ScannerBase#setExecutionHints(Map)}</li>
+ * <li><b>attemptPlans : </b> A list of configuration to use for each scan attempt. Each list object
+ * has the following fields:
  * <ul>
- * <li><b>initialServers</b> the initial number of servers to randomly choose from for a given
- * tablet. Defaults to 3.</li>
- * <li><b>initialBusyTimeout</b>The initial busy timeout to use when contacting a scan servers. If
- * the scan does start running within the busy timeout then another scan server can be tried.
- * Defaults to PT0.033S see {@link Duration#parse(CharSequence)}</li>
- * <li><b>maxBusyTimeout</b>When busy is repeatedly seen, then the busy timeout will be increased
- * exponentially. This setting controls the maximum busyTimeout. Defaults to PT30M</li>
- * <li><b>maxDepth</b>When busy is observed the number of servers to randomly chose from is
- * expanded. This setting controls how many busy observations it will take before we choose from all
- * servers.</li>
+ * <li><b>servers : </b> The number of servers to randomly choose from for this attempt.</li>
+ * <li><b>busyTimeout : </b> The busy timeout to use for this attempt.</li>
+ * <li><b>salt : </b> An optional string to append when hashing the tablet. When this is set
+ * differently for attempts it has the potential to cause the set of servers chosen from to be
+ * disjoint. When not set or the same, the servers between attempts will be subsets.</li>
  * </ul>
+ * </li>
+ * </ul>
+ *
+ * <p>
+ * Below is an example configuration with two profiles, one is the default and the other is used
+ * when the scan execution hint {@code scan_type=slow} is set.
+ * </p>
+ *
+ * <pre>
+ *    [
+ *     {
+ *       "isDefault":true,
+ *       "maxBusyTimeout":"5m",
+ *       "busyTimeoutMultiplier":4,
+ *       "attemptPlans":[
+ *         {"servers":"3", "busyTimeout":"33ms"},
+ *         {"servers":"100%", "busyTimeout":"100ms"}
+ *       ]
+ *     },
+ *     {
+ *       "scanTypeActivations":["slow"]
+ *       "maxBusyTimeout":"20m",
+ *       "busyTimeoutMultiplier":8,
+ *       "attemptPlans":[
+ *         {"servers":"1", "busyTimeout":"10s"},
+ *         {"servers":"3", "busyTimeout":"30s","salt","42"},
+ *         {"servers":"9", "busyTimeout":"60s","salt","84"}
+ *       ]
+ *     }
+ *    ]
+ * </pre>
```

Comment:
I think it might be confusing to have the execution hint key different than the configuration name. In the example above, the execution hint key is `scan_type`, but that does not occur anywhere in the json. I'm wondering if:

  1. The execution hint key should be `profile_name`
  2. `isDefault` and `scanTypeActiviations` in the json should be renamed to `profile_name`
  3. The parsing code should log a msg if the user does not provide their own `default` profile configuration

Reply:
For all of the default plugins that I have written related to scan (like [HintScanPrioritizer](https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/core/spi/scan/HintScanPrioritizer.java) and [SimpleScanDispatcher](https://github.com/apache/accumulo/blob/main/core/src/main/java/org/apache/accumulo/core/spi/scan/SimpleScanDispatcher.java) ) I have made them all be able to respond to the hint `scan_type` via configuration.  My thinking was that users could define scan_types that mean something to them like `gold`,`silver`,`copper` for example and set these when writing code.  Then at runtime they can configure different plugins to respond to these type in different way.  For w/ existing plugins user could do something like the following via runtime configuration with the current builtin plugins.

 * When scan type gold is seen enable full caching and set the scan priority to 1.
 * When scan type silver is seen enable opportunistic caching and set the scan priority to 2.
 * When scan type copper is seen disable caching and send the scan to a thread pool with a single thread.

So I was trying to carry forward the usage of the scan_type scan hint for this new plugin following what the existing plugins have done so that user could do even more configuration based on scan_type hints.  Multiple plugins can respond to the scan type hint, but no individual plugin is obligated to respond to it.  Over time users can change runtime config to respond to the same scan_type in different ways while the system is running.  My overall goal w/ scan_type hints was to avoid putting direct configuration in the scan type hints themselves, but introduce a level of indirection at runtime that I thought gave greater flexibility. 

I was also thinking for this plugin that scanTypeAcitvations could list multiple things like `gold` and `silver` so that a profile could be activated for multiple scan types to avoid duplication when you want to do the same thing for multiple types.

> the execution hint key is scan_type, but that does not occur anywhere in the json. I'm wondering if

There is `scanTypeActivations` in the json.  Its not exactly the same, its camel case but it does mention scan type.

> The parsing code should log a msg if the user does not provide their own default profile configuration

I think it throws an exception if there is not one and only one default profile.

Reply:
Interesting... Might be useful to have an example that incorporates all of the plugins.

Reply:
yeah would probably need to include something in the user manual.

I settled on this pattern when trying to scan hints in Fluo for notification scans.  These are scan where fluo looks for work.  They are low priority full table scans that I would like to limit their core usage and cache usage.  At first I tried passing config from Fluo via scan execution hints and found this to be extremely cumbersome.  Then I settled on passing the scan_type hint 

https://github.com/apache/fluo/blob/5f1ee51f1cfb08eec4b895d40d54984b23b57e10/modules/core/src/main/java/org/apache/fluo/core/worker/finder/hash/ScanTask.java#L197

where I could just change Accumulo config based on that type hint and it was much easier to experiment and configure at runtime via changing Accumulo plugin config (never had to change Fluo code after that.